lib/utils/playlistGenrator-utils.js

```diff
@@ -0,0 +1,1104 @@
+        logger.warn("Playlist.validate !that.durations || typeof that.durations !== 'Array'");
+        return false;
+    }
+
+    if( !_.every(that.durations,function(d){
+            return d >= 0;
+        })  )  {
+        logger.warn("Playlist.validate that.duration < 0");
+        return false;
+    }
+
+    if( !Array.isArray(that.sequences)   ) {
+        logger.warn("Playlist.validate !that.sequences || typeof that.sequences !== 'Array'");
+        return false;
+    }
+
+    for( var i in that.sequences) {
+        var seq = that.sequences[i];
+        if( !Array.isArray(seq.clips) ) {
+            logger.warn("Playlist.validate !seq.clips !seq.clips instanceof Array ");
+            return false;
+        }
+        if( seq.validate && !seq.validate(opts) ){
+            return false;
+        }
+    }
+
+    if( !_.every(that.sequences,function(seq){
+            for( var i in seq.clips ) {
+                var clip = seq.clips[i];
+                var clipD = clip.getTotalDuration(),
+                    overallD = that.durations[i];
+                if(  clipD < overallD ){
+                    return false;
+                }
+            }
+            return true;
+        } ) ) {
+        logger.warn("Playlist.validate duration is greater than the sum of the durations array ");
+        return false;
+    }
+
+    if(that.firstClipTime <  that.segmentBaseTime)  {
+        logger.warn("Playlist.validate that.firstClipTime <  that.segmentBaseTime");
+        return false;
+    }
+
+    return true;
+};
+
+Playlist.prototype.recalculateOffsetsAndDuration = function recalculateOffsetsAndDuration (index){
+    var that = this;
+
+    var minDTS = Number.MIN_VALUE,
+        maxDTS = Number.MAX_VALUE;
+
+    // determine dts range
+    _.each(that.sequences,function(seq){
+        if(seq.clips.length > index) {
+            var dtsRange = seq.clips[index].getDTSRange();
+            minDTS = Math.max(dtsRange.min, minDTS);
+            maxDTS = Math.min(maxDTS, dtsRange.max);
+        }
+    });
+    if(minDTS > Number.MIN_VALUE) {
+
+        that.firstClipTime = Math.max(that.firstClipTime,Math.ceil(minDTS));
+        that.durations[index] =  Math.ceil(maxDTS-that.firstClipTime);
+
+        var date = new Date(that.firstClipTime);
+        logger.info("Playlist.recalculateOffsetsAndDuration set firstClipTime=",date);
+
+        _.each(that.sequences,function(seq) {
+            seq.clips.forEach(function(c) {
+                c.updateOffset(that.firstClipTime);
+            });
+        });
+        return {min:minDTS,max:maxDTS};
+    }
+};
+
+Playlist.prototype.onSerialize = function () {
+    var that = this;
+
+    _.each(that.sequences,function (seq) {
+        seq.clips = seq.clips.map( function(c) {
+            return unserializeClip(c);
+        });
+    });
+};
+
+Playlist.prototype.addListener = function(listener){
+    var that = this;
+
+    BroadcastEventEmitter.prototype.addListener.call(this,listener);
```

Comment:
you should derive from EventEmitter and remove the addListenr and removeListener